### PR TITLE
[Completed][pytorch] add DCGAN

### LINKED ISSUE #1307: [pytorch] code for 17.2.1: The Pokemon Dataset
http://preview.d2l.ai/d2l-en/master/chapter_generative-adversarial-networks/dcgan.html
**My Right Code**: https://github.com/StevenJokes/d2l-en-read/blob/master/chapter-generative-adversarial-networks/dcgan_torch5.ipynb
All my research process: https://github.com/StevenJokes/d2l-en-read/blob/master/chapter-generative-adversarial-networks/dcgan_torch.ipynb
Please help me review the code, and you can continue to translate the behind code.
@AnirudhDagar 

---

**And check [pytorch] code for 17.1 : #1300**:https://github.com/StevenJokes/d2l-en-read/blob/master/chapter-generative-adversarial-networks/my5gan.ipynb
All my research process: https://github.com/StevenJokes/d2l-en-read/blob/master/chapter-generative-adversarial-networks/gan_torch.ipynb

### PR DESCRIPTION
**Pull request by StevenJokes**
*Description of changes:*

By submitting this pull request, I confirm that you can use, modify,
copy, and redistribute this contribution, under the terms of your
choice.

### CODE REVIEW

**Comment by mli** [comment]
Job d2l-en/PR-1309/1 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1309/

**Comment by StevenJokes** [comment]
Thanks. I'll try more.

**Comment by mli** [comment]
Job d2l-en/PR-1309/2 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1309/

**Comment by StevenJokes** [comment]
https://github.com/StevenJokes/d2l-en-read/blob/master/chapter-generative-adversarial-networks/colab_pt_DCGAN/dcgan_pt5.ipynb

I'm stucking about how to train.
Can anyone help me?

**Comment by StevenJokes** [comment]
I can't figure out `update_D` or `update_G` for cpu or gpu.
So I just separate.
https://github.com/StevenJokes/d2l-en-read/blob/master/chapter-generative-adversarial-networks/dcgan_n2train.ipynb
And I can't get the result of train in colab, so just try PR now!

**Comment by StevenJokes** [comment]
What's problem about this check?

**Comment by StevenJokes** [comment]
Thanks for https://discuss.d2l.ai/t/deep-convolutional-generative-adversarial-networks/409/12
@yoderj 
You should be a contributor.

**Comment by StevenJokes** [comment]
My code https://github.com/StevenJokes/d2l-en-read/blob/master/chapter-generative-adversarial-networks/colab_pt_DCGAN/DCGAN_ptlr5.ipynb

**Comment by mli** [comment]
Job d2l-en/PR-1309/9 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1309/

**Comment by mli** [comment]
Job d2l-en/PR-1309/10 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1309/

**Comment by StevenJokes** [comment]
Please review?
@AnirudhDagar

File: chapter_generative-adversarial-networks/dcgan.md [hunk 1]
```diff
@@ -5,19 +5,28 @@ In :numref:`sec_basic_gan`, we introduced the basic ideas behind how GANs work.
...
 npx.set_np()
 ```
 
+```{.python .input}
+#@tab pytorch
+from d2l import torch as d2l
+import torch
+import torchvision
+from torch import nn
+import warnings
```
Comment (AnirudhDagar) [inline chapter_generative-adversarial-networks/dcgan.md:22]: We don't filter out warnings. You need to solve these warnings.

File: chapter_generative-adversarial-networks/dcgan.md [hunk 2]
```diff
@@ -40,21 +59,45 @@ data_iter = gluon.data.DataLoader(
...
     break
 ```
 
+```{.python .input}
+#@tab pytorch
+warnings.filterwarnings('ignore')
+d2l.set_figsize((4, 4))
+for X, y in data_iter:
+    imgs = X[0:20,:,:,:].permute(0, 2, 3, 1)/2+0.5
+    d2l.show_images(imgs, num_rows=4, num_cols=5)
```
Comment (AnirudhDagar) [inline chapter_generative-adversarial-networks/dcgan.md:92]: Don't filter warnings. Solve the warning. Maybe don't use `permute` to achieve the same output without warnings.

File: chapter_generative-adversarial-networks/dcgan.md [hunk 3]
```diff
@@ -109,14 +181,36 @@ net_G.add(G_block(n_G*8, strides=1, padding=0),  # Output: (64 * 8, 4, 4)
               activation='tanh'))  # Output: (3, 64, 64)
 ```
 
+```{.python .input}
+#@tab pytorch
+def Conv2DTranspose(channels, kernel_size, strides, padding, use_bias, nc=3):
+    return nn.ConvTranspose2d(nc, channels, kernel_size=kernel_size,stride=strides, padding=padding, bias=use_bias)
```
Comment (AnirudhDagar) [inline chapter_generative-adversarial-networks/dcgan.md:187]: What is the need to define this function?

File: chapter_generative-adversarial-networks/dcgan.md [hunk 4]
```diff
@@ -125,16 +219,25 @@ $$\textrm{leaky ReLU}(x) = \begin{cases}x & \text{if}\ x > 0\\ \alpha x &\text{o
...
-d2l.plot(x.asnumpy(), Y, 'x', 'y', alphas)
+d2l.plot(d2l.numpy(x), Y, 'x', 'y', alphas)
+```
+
+```{.python .input}
+#@tab pytorch
+alphas = [0, 0.2, 0.4, .6, .8, 1]
+x = torch.arange(-2, 1, 0.1)
+Y = [nn.LeakyReLU(alpha)(x).numpy() for alpha in alphas]
+d2l.plot(d2l.numpy(x), Y, 'x', 'y', alphas)
```
Comment (AnirudhDagar) [inline chapter_generative-adversarial-networks/dcgan.md:234]: Merge code blocks

File: chapter_generative-adversarial-networks/dcgan.md [hunk 5]
```diff
@@ -177,19 +303,40 @@ net_D.add(D_block(n_D),   # Output: (64, 32, 32)
           nn.Conv2D(1, kernel_size=4, use_bias=False))  # Output: (1, 1, 1)
 ```
 
+```{.python .input}
+#@tab pytorch
+def Conv2D(channels, kernel_size, use_bias, nc=3):
+    return nn.Conv2d(nc, channels, kernel_size=kernel_size, bias=use_bias)
```
Comment (AnirudhDagar) [inline chapter_generative-adversarial-networks/dcgan.md:309]: Why define this?

File: chapter_generative-adversarial-networks/dcgan.md [hunk 6]
```diff
@@ -229,19 +376,93 @@ def train(net_D, net_G, data_iter, num_epochs, lr, latent_dim,
...
+    ones = torch.ones((batch_size,1,1,1), device=Z.device)
+    trainer_G.zero_grad()
+    # We could reuse `fake_X` from `update_D` to save computation
+    fake_X = net_G(Z)
+    # Recomputing `fake_Y` is needed since `net_D` is changed
+    fake_Y = net_D(fake_X)
+    loss_G = loss(fake_Y, ones)
+    loss_G.backward()
+    trainer_G.step()
+    return loss_G
```
Comment (AnirudhDagar) [inline chapter_generative-adversarial-networks/dcgan.md:409]: Use d2l.update... like the mxnet version

File: chapter_generative-adversarial-networks/dcgan.md [hunk 7]
```diff
@@ -229,19 +376,93 @@ def train(net_D, net_G, data_iter, num_epochs, lr, latent_dim,
...
+            batch_size = X.shape[0]
+            Z = torch.normal(0, 1, size=(batch_size, latent_dim, 1, 1))
+            X, Z = X.to(device), Z.to(device)
+            metric.add(update_D(X, Z, net_D, net_G, loss, trainer_D),
+                       update_G(Z, net_D, net_G, loss, trainer_G),
+                       batch_size)
+        # Show generated examples
+        Z = torch.normal(0, 1, size=(21, latent_dim, 1, 1), device=device)
+        # Normalize the synthetic data to N(0, 1)
+        fake_x = net_G(Z).permute(0, 2, 3, 1) / 2 + 0.5
```
Comment (AnirudhDagar) [inline chapter_generative-adversarial-networks/dcgan.md:439]: Don't use permute. Use reshape or some other tensor manipulation.

**Review by AnirudhDagar — CHANGES_REQUESTED** [review]

File: chapter_generative-adversarial-networks/dcgan.md (hunk 6 above)
Comment (StevenJokes) [reply chapter_generative-adversarial-networks/dcgan.md:409]: torch.ones((batch_size, 1, 1, 1)
I can't. The num of dim isn't same.

File: chapter_generative-adversarial-networks/dcgan.md (hunk 5 above)
Comment (StevenJokes) [reply chapter_generative-adversarial-networks/dcgan.md:309]: To look similar to mxnet.

File: chapter_generative-adversarial-networks/dcgan.md (hunk 3 above)
Comment (StevenJokes) [reply chapter_generative-adversarial-networks/dcgan.md:187]: To look similar to mxnet.

File: chapter_generative-adversarial-networks/dcgan.md (hunk 6 above)
Comment (StevenJokes) [reply chapter_generative-adversarial-networks/dcgan.md:409]: .ones((batch_size,))
mxnet can add dims auto.
But pytorch need to require dim advancely.

**Comment by StevenJokes** [comment]
image
It seems to happen in using `for`, but not in using `permute`

File: chapter_generative-adversarial-networks/dcgan.md (hunk 5 above)
Comment (StevenJokes) [reply chapter_generative-adversarial-networks/dcgan.md:309]: we need to know `nc=3`

**Comment by StevenJokes** [comment]
image

File: chapter_generative-adversarial-networks/dcgan.md (hunk 1 above)
Comment (StevenJokes) [reply chapter_generative-adversarial-networks/dcgan.md:22]: See my comments below.

File: chapter_generative-adversarial-networks/dcgan.md (hunk 6 above)
Comment (StevenJokes) [reply chapter_generative-adversarial-networks/dcgan.md:409]: Look GAN's d2l.update: image

I have tried, but failed.

**Comment by StevenJokes** [comment]
And if I don't filter warnings, the training picture can't be showed until we finish or pause the training cell.
You can try by yourself. Now I can't find a better way.
image
image

File: chapter_generative-adversarial-networks/dcgan.md (hunk 7 above)
Comment (StevenJokes) [reply chapter_generative-adversarial-networks/dcgan.md:439]: `torch.reshape` can only switch two locations, not four
permute is equal to reshape in numpy.

File: chapter_generative-adversarial-networks/dcgan.md (hunk 5 above)
Comment (StevenJokes) [reply chapter_generative-adversarial-networks/dcgan.md:309]: we need to know nc=3

File: chapter_generative-adversarial-networks/dcgan.md (hunk 4 above)
Comment (StevenJokes) [reply chapter_generative-adversarial-networks/dcgan.md:234]: merge now!

File: chapter_generative-adversarial-networks/dcgan.md (hunk 2 above)
Comment (StevenJokes) [reply chapter_generative-adversarial-networks/dcgan.md:92]: See my comments below.

**Comment by mli** [comment]
Job d2l-en/PR-1309/11 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1309/

File: chapter_generative-adversarial-networks/dcgan.md (hunk 3 above)
Comment (AnirudhDagar) [reply chapter_generative-adversarial-networks/dcgan.md:187]: Just use the `nn.ConvTranspose2d` as is. No need to define a separate function with a different name doing exactly that.

Comment (StevenJokes) [reply chapter_generative-adversarial-networks/dcgan.md:187]: we need to know nc=3
I think it will be friendly to new users to show the similiarity and difference of pytorch and mxnet.

Comment (StevenJokes) [reply chapter_generative-adversarial-networks/dcgan.md]: But I followed your advice.
Look my newest PR.

**Comment by mli** [comment]
Job d2l-en/PR-1309/12 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1309/

**Review by AnirudhDagar — COMMENTED** [review]
Never mind. I'm merging this and will make a few fixes.
Thanks! :)
